# This is a collection of some of my work as a showcase to prospective employers
I have collected some of the work i have created during my studies and private time, If you are a prospective employer looking here's a brief overview of what each folder means

## AI_practice
I created a AI model to classify retinal scans into 3 different classes (normal,glaucoma and cataracts), 
i trained and tested it using Stratfied K fold and created different visualisations, I have included both the ipynb and a PDF which contains the results

**Demonstrates** :
- python
- Tensforflow
- Machine learning principles (building,training,testing and evaluating models)


## scRNA
scRNA (single cell RNA), i used both Seurat and Scanpy. I used some Rhesus monkey single cell data to analyse.

**Demonstrates** :
- Seurat
- Scanpy

### data folder
this contains the actual data for the Rhesus monkeys

## database and data management
Here i show some of my skill in database creation, this was for a project during my course. Where we had to take data, clean it and make it into a database (and R shiny app)
I personally created the database used R to clean the  raw data to get it into the right format to upload onto the database.

**Demonstrates** :
- SQL
- mySQL
- R for data cleaning

## Cleaning_classification_clustering
This was done during my reseach project, i had to classify cell lines into 2 extreme phenotypes based on some different paramaters
I used both logarithmic and logistic models to fit the data, as well as some regression. I also used PCA and K means clustering to sort out.

**Demonstrates** :
1. Mathematical modelling
2. Data cleaning
3. Clustering and Principle component use
4. Biological understanding

## Sequence_alignment_and analysis
In this i recreate the results from this paper https://pubs.acs.org/doi/10.1021/acssynbio.9b00027. Listed is the sequence alignment using the cufflinks suite. T
here were approx 40 samples FASTA files ,that needed to be downloaded , cleaned and QC'd before going into sequence alignment. This script was ran on a Slurm HPC.
This was then fed into cuffdiff for DE_analysis. CummeRbund to analyse and reproduce results.

**Demonstrates** :
1. Sequence Alignemnt (HISAT2)
2. Differential expression Analsysis (Cufflinks)
3. Shell scripting
4. Visualisation
5. Ability to read and understand and recreate works
6. Containerisation (singularity

## skills not shown explicitily here
here are some things i'm working on but are not yet ready to be pushed to the online repository
- Variant calling and annotation
- Workflow Software (nextflow) 
- Docker

## Further info
please feel free to contact me for some more information on these projects (including the reports, academic grades, etc)

